File: backend/src/main/java/com/soccerdreamfermana/service/DonationService.java
package com.soccerdreamfermana.service;

import com.soccerdreamfermana.dto.donation.CapturePaymentResponse;
import com.soccerdreamfermana.dto.donation.CreatePaymentRequest;
import com.soccerdreamfermana.dto.donation.CreatePaymentResponse;
import com.soccerdreamfermana.dto.donation.DonationRequest;
import com.soccerdreamfermana.dto.donation.DonationResponse;
import com.soccerdreamfermana.exception.BadRequestException;
import com.soccerdreamfermana.exception.ResourceNotFoundException;
import com.soccerdreamfermana.model.Donation;
import com.soccerdreamfermana.payment.PaymentAdapter;
import com.soccerdreamfermana.repository.DonationRepository;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class DonationService {

 private static final String DONATION_NOT_FOUND_MESSAGE = "Donazione non trovata";
 private static final String INVALID_PAYMENT_STATUS_MESSAGE = "La donazione può essere salvata solo dopo una cattura completata";
 private static final String COMPLETED_PAYMENT_STATUS = "COMPLETED";

 private final DonationRepository donationRepository;
 private final PaymentAdapter paymentAdapter;

 public CreatePaymentResponse createPayment(CreatePaymentRequest request) {
  return paymentAdapter.createPayment(request);
 }

 public CapturePaymentResponse capturePayment(String orderId) {
  return paymentAdapter.capturePayment(orderId);
 }

 public DonationResponse create(DonationRequest request) {
  validatePaymentStatus(request.paymentStatus());
  return toResponse(donationRepository.save(buildDonation(request)));
 }

 public List<DonationResponse> list() {
  return donationRepository.findAllByOrderByCreatedAtDesc().stream().map(this::toResponse).toList();
 }

 public DonationResponse getById(Long id) {
  return toResponse(findDonationById(id));
 }

 public void delete(Long id) {
  donationRepository.delete(findDonationById(id));
 }

 private void validatePaymentStatus(String paymentStatus) {
  if (paymentStatus != null && !COMPLETED_PAYMENT_STATUS.equalsIgnoreCase(paymentStatus)) {
   throw new BadRequestException(INVALID_PAYMENT_STATUS_MESSAGE);
  }
 }

 private Donation buildDonation(DonationRequest request) {
  return Donation.builder()
   .nome(request.nome().trim())
   .email(normalizeEmail(request.email()))
   .importo(request.importo())
   .paypalOrderId(request.paypalOrderId())
   .payerId(request.payerId())
   .captureId(request.captureId())
   .paymentStatus(resolvePaymentStatus(request.paymentStatus()))
   .build();
 }

 private Donation findDonationById(Long id) {
  return donationRepository.findById(id).orElseThrow(() -> new ResourceNotFoundException(DONATION_NOT_FOUND_MESSAGE));
 }

 private DonationResponse toResponse(Donation donation) {
  return new DonationResponse(
   donation.getId(),
   donation.getNome(),
   donation.getEmail(),
   donation.getImporto(),
   donation.getPaypalOrderId(),
   donation.getPayerId(),
   donation.getCaptureId(),
   donation.getPaymentStatus(),
   donation.getCreatedAt()
  );
 }

 private String normalizeEmail(String email) {
  return email.trim().toLowerCase();
 }

 private String resolvePaymentStatus(String paymentStatus) {
  return paymentStatus == null ? COMPLETED_PAYMENT_STATUS : paymentStatus;
 }
}

File: backend/Services/DonationService.cs
using Microsoft.EntityFrameworkCore;
using SoccerDreamFermana.Backend.Data;
using SoccerDreamFermana.Backend.Dtos;
using SoccerDreamFermana.Backend.Exceptions;
using SoccerDreamFermana.Backend.Models;
using SoccerDreamFermana.Backend.Payments;

namespace SoccerDreamFermana.Backend.Services;

public sealed class DonationService(AppDbContext db, IPaymentAdapter paymentAdapter)
{
    private const string DonationNotFoundMessage = "Donazione non trovata";
    private const string InvalidPaymentStatusMessage = "La donazione puo essere salvata solo dopo una cattura completata";
    private const string CompletedPaymentStatus = "COMPLETED";

    public Task<CreatePaymentResponse> CreatePaymentAsync(CreatePaymentRequest request, CancellationToken cancellationToken)
    {
        return paymentAdapter.CreatePaymentAsync(request, cancellationToken);
    }

    public Task<CapturePaymentResponse> CapturePaymentAsync(string orderId, CancellationToken cancellationToken)
    {
        return paymentAdapter.CapturePaymentAsync(orderId, cancellationToken);
    }

    public async Task<DonationResponse> CreateAsync(DonationRequest request, CancellationToken cancellationToken)
    {
        ValidatePaymentStatus(request.PaymentStatus);
        var donation = new Donation
        {
            Nome = TextNormalizer.NormalizeRequired(request.Nome),
            Email = TextNormalizer.NormalizeEmail(request.Email),
            Importo = request.Importo,
            PaypalOrderId = request.PaypalOrderId,
            PayerId = request.PayerId,
            CaptureId = request.CaptureId,
            PaymentStatus = string.IsNullOrWhiteSpace(request.PaymentStatus) ? CompletedPaymentStatus : request.PaymentStatus
        };

        db.Donations.Add(donation);
        await db.SaveChangesAsync(cancellationToken);
        return ToResponse(donation);
    }

    public async Task<IReadOnlyList<DonationResponse>> ListAsync(CancellationToken cancellationToken)
    {
        var donations = await db.Donations
            .AsNoTracking()
            .OrderByDescending(donation => donation.CreatedAt)
            .ToListAsync(cancellationToken);

        return donations.Select(ToResponse).ToList();
    }

    public async Task<DonationResponse> GetByIdAsync(long id, CancellationToken cancellationToken)
    {
        return ToResponse(await FindDonationByIdAsync(id, cancellationToken));
    }

    public async Task DeleteAsync(long id, CancellationToken cancellationToken)
    {
        db.Donations.Remove(await FindDonationByIdAsync(id, cancellationToken));
        await db.SaveChangesAsync(cancellationToken);
    }

    private static void ValidatePaymentStatus(string? paymentStatus)
    {
        if (paymentStatus is not null && !CompletedPaymentStatus.Equals(paymentStatus, StringComparison.OrdinalIgnoreCase))
        {
            throw new BadRequestException(InvalidPaymentStatusMessage);
        }
    }

    private async Task<Donation> FindDonationByIdAsync(long id, CancellationToken cancellationToken)
    {
        return await db.Donations.FindAsync(new object?[] { id }, cancellationToken)
            ?? throw new ResourceNotFoundException(DonationNotFoundMessage);
    }

    private static DonationResponse ToResponse(Donation donation)
    {
        return new DonationResponse(
            donation.Id,
            donation.Nome,
            donation.Email,
            donation.Importo,
            donation.PaypalOrderId,
            donation.PayerId,
            donation.CaptureId,
            donation.PaymentStatus,
            donation.CreatedAt
        );
    }
}
